src/style-spec/feature_filter/convert.js

```diff
@@ -0,0 +1,197 @@
+// @flow
+
+import { isExpressionFilter } from './index';
+
+import type { FilterSpecification } from '../types';
+
+type ExpectedTypes = {[string]: 'string' | 'number' | 'boolean'};
+
+/**
+ * Convert the given legacy filter to (the JSON representation of) an
+ * equivalent expression
+ * @private
+ */
+export default function convertFilter(filter: FilterSpecification): mixed {
+    return _convertFilter(filter, {});
+}
+
+/*
+ * Convert the given filter to an expression, storing the expected types for
+ * any feature properties referenced in expectedTypes.
+ *
+ * These expected types are needed in order to construct preflight type checks
+ * needed for handling 'any' filters. A preflight type check is necessary in
+ * order to mimic legacy filters' semantics around expected type mismatches.
+ * For example, consider the legacy filter:
+ *
+ *     ["any", ["all", [">", "y", 0], [">", "y", 0]], [">", "x", 0]]
+ *
+ * Naively, we might convert this to the expression:
+ *
+ *     ["any", ["all", [">", ["get", "y"], 0], [">", ["get", "z"], 0]], [">", ["get", "x"], 0]]
+ *
+ * But if we tried to evaluate this against, say `{x: 1, y: null, z: 0}`, the
+ * [">", ["get", "y"], 0] would cause an evaluation error, leading to the
+ * entire filter returning false. Legacy filter semantics, though, ask for
+ * [">", "y", 0] to simply return `false` when `y` is of the wrong type,
+ * allowing the subsequent terms of the outer "any" expression to be evaluated
+ * (resulting, in this case, in a `true` value, because x > 0).
+ *
+ * We account for this by inserting a preflight type-checking expression before
```

Comment:
Can you add a few sentences outlining the motivation for this particular approach as opposed to alternatives? I'm assuming it's to minimize the cases where typechecking is needed (only when "any" is used; other cases are fine to let error out [Are they? Will this produce annoying error messages?]).

Reply:
> I'm assuming it's to minimize the cases where typechecking is needed (only when "any" is used; other cases are fine to let error out [Are they? Will this produce annoying error messages?])

Yep, that was the reason.  Other cases are fine to let error out, in the sense that they'll produce the correct behavior.  It's true that they'll produce annoying warnings in that case. But this (and the other issues that have come up re: evaluation error warnings) makes me wonder whether it might make sense to simply disable these warnings (or some subset of them?) by default, only showing them if a `map.expressionDebug` property is set to true...

Reply:
> simply disable these warnings (or some subset of them?)

Another possibility: we currently only log each distinct evaluation error once, with that uniqueness scoped to each expression. If we expanded the uniqueness scope to the whole map, that would dramatically reduce the number of these warnings, while still giving developers a clue if they're trying to figure out why their style isn't doing what they want.